.github/workflows/main.yml

```diff
@@ -2,66 +2,193 @@ name: main
               run: yarn
 
-            - name: Format code
-              run: yarn format
-    tests:
+    changed-files:
         runs-on: ubuntu-latest
-
+        outputs:
+            any_changed: ${{ steps.changed-files.outputs.any_changed }}
+            modified_files: ${{ steps.changed-files.outputs.modified_files }}
         steps:
-            - name: Checkout
-              uses: actions/checkout@v4
+            - uses: actions/checkout@v4
+            - name: Get changed files
+              id: changed-files
+              uses: tj-actions/changed-files@v44
+              with:
+                  files: packages/**/*.{sol,json,ts}
 
-            - name: Setup Node
-              uses: actions/setup-node@v4
+    compile:
+        if: needs.changed-files.outputs.any_changed == 'true'
+        needs: [changed-files, deps]
+        runs-on: ubuntu-latest
+        steps:
+            - uses: actions/checkout@v4
+            - uses: actions/setup-node@v4
               with:
-                  node-version: "20"
-                  cache: yarn
+                  node-version: 20
+            - uses: actions/cache/restore@v4
+              with:
+                  path: node_modules
+                  key: ${{ needs.deps.outputs.cache-key }}
 
-            - name: Install dependencies
-              run: yarn
+            - run: yarn compile
 
-            - name: Get changed files
-              id: changed-files
-              uses: tj-actions/changed-files@v44
+            - name: Upload compilation results
+              uses: actions/upload-artifact@v4
               with:
-                  files: |
-                      packages/**/*.{sol,json,ts}
+                  name: all-artifacts
+                  path: packages/**/artifacts/**
 
-            - if: steps.changed-files.outputs.any_changed == 'true'
-              name: Compile contracts
-              run: yarn compile
+    style:
+        needs: deps
+        runs-on: ubuntu-latest
+        steps:
+            - uses: actions/checkout@v4
+            - uses: actions/setup-node@v4
+              with:
+                  node-version: 20
+            - uses: actions/cache/restore@v4
+              with:
+                  path: node_modules
+                  key: ${{ needs.deps.outputs.cache-key }}
+
+            - run: yarn format
+    tests:
+        if: needs.changed-files.outputs.any_changed == 'true'
+        needs: [changed-files, set-matrix, deps, compile]
+        runs-on: ubuntu-latest
+        strategy:
+            matrix:
+                dir: ${{ fromJson(needs.set-matrix.outputs.matrix) }}
+
+        steps:
+            - uses: actions/checkout@v4
+            - uses: actions/setup-node@v4
+              with:
+                  node-version: 20
+            - uses: actions/cache/restore@v4
+              with:
+                  path: node_modules
+                  key: ${{ needs.deps.outputs.cache-key }}
+            - uses: actions/download-artifact@v4
```

Comment:
This is basically getting the artifacts already generated in the other job (i.e. compile) right?

Reply:
exactly